src/marin/processing/classification/inference.py

```diff
@@ -26,215 +26,424 @@
+            elif col in batch_dict:
+                output_row[col] = batch_dict[col][i]
+        output_rows.append(output_row)
 
+    return output_rows
 
-@cached_or_construct_output(success_suffix="SUCCESS")
-def process_file_with_quality_classifier(input_filename: str, output_filename: str, quality_classifier: BaseClassifier):
-    print(f"[*] Processing {input_filename} to {output_filename}")
 
-    dataset = read_dataset(input_filename)
+def write_dataset_streaming(rows_iterator, output_filename: str, append: bool = False):
+    """Writes rows to a file in streaming fashion
 
-    # TODO(chris): Add support for more types of columns.
-    input_column_names = get_input_dataset_column_names(input_filename)
-    output_column_names = get_output_dataset_column_names(input_filename)
-    dataset = dataset.select_columns(input_column_names)
-    dataset = dataset.map(lambda batch: quality_classifier(batch), batched=True, batch_size=512)
-    dataset = dataset.select_columns(output_column_names)
+    JSONL behavior:
+      - For JSONL(.gz/.zst), write to a deterministic temp file under /tmp in append-mode,
+        then upload the full file to the destination (e.g., gs://...). This provides safe
+        append semantics for object stores.
+      - Checkpoint restoration should continue to read from the remote path.
+    """
+    import json
+    import hashlib
+    import os
+    import shutil
+
+    import fsspec
+
+    mode = "ab" if append else "wb"
+
+    if ".jsonl" in output_filename:
+        # Build deterministic temp path for this output file
+        file_hash = hashlib.sha256(output_filename.encode("utf-8")).hexdigest()
+        if output_filename.endswith(".jsonl.gz"):
+            tmp_path = f"/tmp/marin_{file_hash}.jsonl.gz"
+        elif output_filename.endswith(".jsonl.zst"):
+            tmp_path = f"/tmp/marin_{file_hash}.jsonl.zst"
+        else:
+            tmp_path = f"/tmp/marin_{file_hash}.jsonl"
+
+        # If appending and local temp doesn't exist, hydrate it from remote (if present)
+        if append and not os.path.exists(tmp_path):
+            fs, _ = fsspec.core.url_to_fs(output_filename)
+            if fs.exists(output_filename):
+                with fsspec.open(output_filename, "rb") as src, open(tmp_path, "wb") as dst:
+                    shutil.copyfileobj(src, dst)
+            else:
+                # Ensure file exists for append
+                open(tmp_path, "wb").close()
+
+        # Write rows to temp with append semantics using fsspec with compression inference
+        with fsspec.open(tmp_path, mode, compression="infer") as f:
+            for row in rows_iterator:
+                row = make_json_serializable(row)
+                f.write((json.dumps(row) + "\n").encode("utf-8"))
+
+        # Upload temp file to destination (overwrite remote with full content)
+        with fsspec.open(output_filename, "wb") as dst, open(tmp_path, "rb") as src:
+            shutil.copyfileobj(src, dst)
```

Comment:
nit: 

consider returning after each branch, and then the fallthrough is just raise ValueError()